ml-proto/host/parser.mly

```diff
@@ -238,35 +284,49 @@ cases :
 func_fields :
   | expr_list
     { let at = at () in
-      fun c ->
-      {params = []; result = None; locals = [];
-       body = Sugar.func_body ($1 c) @@ at} }
+      empty_type,
+      fun c -> let body = Sugar.func_body ($1 c) @@ at in
+        {ftype = -1 @@ at; locals = []; body} }
   | LPAR PARAM value_type_list RPAR func_fields
-    { fun c -> anon_locals c $3; let f = $5 c in
-      {f with params = $3 @ f.params} }
-  | LPAR PARAM bind_var value_type RPAR func_fields  /* Sugar */
-    { fun c -> bind_local c $3; let f = $6 c in
-      {f with params = $4 :: f.params} }
-  | LPAR RESULT value_type RPAR func_fields
-    { let at = at () in
-      fun c -> let f = $5 c in
-      match f.result with
-      | Some _ -> Error.error at "more than one return type"
-      | None -> {f with result = Some $3} }
+    { {(fst $5) with ins = $3 @ (fst $5).ins},
+      fun c -> anon_locals c $3; (snd $5) c }
```

Comment:
Just a note to reviewers: so the return value of `func_fields` is now a pair, instead of just the function.  This was necessary to extract the type during the first pass (when the global scope is built) instead of the second pass (where function bodies are generated).  Alternatively, I played a bit with moving the param/result fields out of `func_fields` and into another production that came before `func_fields` in `func`, but I ran into shift/reduce and reduce/reduce conflicts that seemed to require changing the text format to better isolate the signature.